ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceBlobSidecarsAvailabilityChecker.java

```diff
@@ -109,12 +149,212 @@ private BlobSidecarsAndValidationResult internalValidate(final List<BlobSidecar>
+          && !isBlockInDataAvailabilityWindow()) {
+        // there are no available blobs so far, but we are outside the availability window. We can
+        // skip additional checks
+        return Optional.of(BlobSidecarsAndValidationResult.NOT_REQUIRED);
+      }
+    }
+
+    // perform the actual validation
+    final BlobSidecarsAndValidationResult result =
+        internalValidate(blobSidecarsToValidate, kzgCommitmentsToValidate);
+
+    if (result.isFailure()) {
+      return Optional.of(result);
+    }
+
+    if (performCompleteValidation) {
+      return Optional.of(result);
+    }
+
+    // cache partially validated blobs
+    blobSidecarsToValidate.forEach(
+        blobSidecar -> validatedBlobSidecars.put(blobSidecar.getIndex(), blobSidecar));
+
+    return Optional.empty();
+  }
+
+  /**
+   * Step 2 of 2
+   *
+   * <p>This function completes the validation if necessary. If there are still blobs to validate,
+   * it waits for the tracker to complete and then performs the final validation by computing and
+   * validating the non-yet validated blobs-commitments, and then computes the final validation
+   * result with the entire list of blobs
+   *
+   * @param maybeBlobSidecarsAndValidationResult is the validation result coming from Step 1.
+   * @return final validation result
+   */
+  private SafeFuture<BlobSidecarsAndValidationResult> completeValidation(
+      final Optional<BlobSidecarsAndValidationResult> maybeBlobSidecarsAndValidationResult) {
+
+    return maybeBlobSidecarsAndValidationResult
+        .map(SafeFuture::completedFuture)
+        .orElseGet(
+            () ->
+                blockBlobSidecarsTracker
+                    .getCompletionFuture()
+                    .orTimeout(waitForTrackerCompletionTimeout)
+                    .thenApplyChecked(__ -> internalValidateAdditional())
+                    .thenApply(this::internalComputeFinalValidationResult)
+                    .exceptionallyCompose(
+                        error ->
+                            ExceptionUtil.getCause(error, TimeoutException.class)
+                                .map(
+                                    timeoutException ->
+                                        SafeFuture.completedFuture(
+                                            BlobSidecarsAndValidationResult.notAvailable(
+                                                timeoutException)))
+                                .orElseGet(() -> SafeFuture.failedFuture(error))));
+  }
+
+  /**
+   * Knowing that:
+   *
+   * <p>(1) `validatedBlobSidecars` contains already validated blobsSidecars
+   *
+   * <p>(2) `blockBlobSidecarsTracker` is now completed
+   *
+   * <p>This function computes and validates a batch of BlobSidecars-kzgCommitments that needs to be
+   * additionally validated to verify the entire set
+   *
+   * @return validation result for the batch
+   */
+  private BlobSidecarsAndValidationResult internalValidateAdditional() {
```

Comment:
Maybe use word `remaining`? Additional confuses with additional checks rather than additional blobSidecars for me